File: kotlin/src/mal/step1_read_print.kt
fun READ(s: String) = read_str(s)

fun EVAL(s: MalType) = s

fun PRINT(v: MalType) = pr_str(v)

fun rep(s: String) {
    println(PRINT(EVAL(READ(s))))
}

fun main(args: Array<String>) {
    while(true) {
        print("user> ")

        try {
            readLine()?.let { rep(it) }
        }
        catch(e: Exception) {
            println("Oh dear:" + e.toString())
        }
    }
}

File: impls/erlang/src/step1_read_print.erl
-module(step1_read_print).

-export([main/1]).

main(_) ->
    case io:get_line(standard_io, "user> ") of
        eof ->
            io:format("~n"),
            ok;
        {error, ErrorDescription} ->
            io:format("Error reading input: ~p~n", [ErrorDescription]),
            exit(ioerr);
        Input ->
            io:format("~s~n", [print(eval(read(Input)))]),
            main("")
        end.

read(Code) ->
    % TODO Errors!
    reader:read_str(Code).

eval(Code) ->
    Code.

print(Ast) ->
    printer:pr_str(Ast).

